tests/cpp/test_multidevice_pipeline.cpp

```diff
@@ -576,6 +581,304 @@ INSTANTIATE_TEST_SUITE_P(
         SchedulingMode::Manual,
         SchedulingMode::ReductionOnly,
         SchedulingMode::Automatic));
+
+class DistributedMatmul : public MultiDeviceTest {
+ protected:
+  DistributedMatmul() : optimization_guard_(false) {
+    DisableOptionsGuard::getCurOptions().set(DisableOption::MatmulExprEval);
+  }
+
+  void SetUp() {
+    if (!deviceMajorMinorCheck(8)) {
+      GTEST_SKIP() << "Distributed matmul tests require Ampere or newer";
+    }
+    MultiDeviceTest::SetUp();
+    num_devices = communicator->size();
+  }
+
+  // use_fusion_executor cache=true, skip_auto_scheduling=false
+  // cache_fusion_executor=false
+  MultiDeviceExecutorParams executor_params{true, false, false};
+  int num_devices;
+
+  DeviceMesh createDeviceMesh() {
+    std::vector<int64_t> devices(num_devices);
+    std::iota(devices.begin(), devices.end(), 0);
+    return DeviceMesh(devices);
+  }
+
+  ValidationConstants getTolerances() {
+    ValidationConstants tolerance_overwrite = ValidationConstants();
+    std::array<std::array<double, 2>, 20> relaxed_sum_tol;
+    for (auto& arr : relaxed_sum_tol) {
+      arr = {128, 2e-4};
+    }
+    tolerance_overwrite.sum_tolerances_float = relaxed_sum_tol;
+    return tolerance_overwrite;
+  }
+
+  std::tuple<at::Tensor, at::Tensor, at::Tensor> getAtenInputOutputs(
+      MmaLayout layout,
+      int M,
+      int N,
+      int K) {
+    int device = communicator->deviceId();
+    c10::ScalarType type = c10::ScalarType::Half;
+    auto a =
+        matmulAtInput2D(layout, TensorMatmulPos::A, type, M, N, K, 0, device);
+    auto b =
+        matmulAtInput2D(layout, TensorMatmulPos::B, type, M, N, K, 0, device);
+    auto c =
+        atMatmul(a.to(at::kDouble), b.to(at::kDouble), layout).to(at::kFloat);
+    return std::make_tuple(a, b, c);
+  }
+
+ private:
+  preseg_passes::OptimizationPassGuard<preseg_passes::AllocationDomainPass>
+      optimization_guard_;
+  DisableOptionsGuard option_guard_;
+};
+
+TEST_F(DistributedMatmul, LayoutTN_NoComms) {
```

Comment:
Is it possible to refactor the four different layouts in one test depending on a "layout param"? This would reduce duplication a lot and will be easier to read maintain

Reply:
I would argue that would made the tests harder to read since some of the tests diverge and we would need to add many if statements to check which layout and which shardings.